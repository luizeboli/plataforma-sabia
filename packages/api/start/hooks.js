const { hooks } = require('@adonisjs/ignitor');

const AlgoliaFakeProvider = require('../test/utils/AlgoliaFakeService');

hooks.before.providersRegistered(() => {
	if (process.env.NODE_ENV === 'testing') {
		/*
		|--------------------------------------------------------------------------
		| Add fakes
		|--------------------------------------------------------------------------
		|
		| Fake provider services before instantiation
		|
		*/
		const { ioc } = use('@adonisjs/fold');

		ioc.singletonFake('App/Services/AlgoliaSearch', () => {
			return AlgoliaFakeProvider;
		});
	}
});

hooks.after.providersBooted(() => {
	const Validator = use('Validator');
	const Database = use('Database');

	const existsFn = async (data, field, message, args, get) => {
		const value = get(data, field);
		if (!value) {
			/**
			 * skip validation if value is not defined. `required` rule
			 * should take care of it.
			 */
			return;
		}

		const [table, column] = args;
		const row = await Database.table(table)
			.where(column, value)
			.first();

		if (!row) {
			throw message;
		}
	};

	// eslint-disable-next-line global-require
	const cpf = require('../app/Utils/cpf');
	Validator.extend('exists', existsFn);
	Validator.extend('cpf', cpf);
});
