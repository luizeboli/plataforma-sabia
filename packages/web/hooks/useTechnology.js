import { useContext } from 'react';
import { TechnologyContext } from '../components/Technology';

function useTechnology() {
	return useContext(TechnologyContext);
}

export default useTechnology;
