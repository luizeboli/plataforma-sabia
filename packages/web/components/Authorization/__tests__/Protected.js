/* eslint-disable jsx-a11y/aria-role */
import React from 'react';
import { render } from 'test-utils';
import * as Router from 'next/router';
import { Protected } from '..';
import * as useAuth from '../../../hooks/useAuth';

describe('Protected component', () => {
	it('should return the login modal if user is not logged in', () => {
		const childrenText = 'children';

		const { container, queryByTestId } = render(
			<Protected>
				<h1>{childrenText}</h1>
			</Protected>,
		);

		expect(container).toMatchSnapshot();
		expect(container.querySelector('h1').textContent).not.toEqual(childrenText);
		expect(queryByTestId('modal')).toBeTruthy();
		expect(container.querySelector('form input[type=email]')).toBeTruthy();
		expect(container.querySelector('form input[type=password]')).toBeTruthy();
		expect(container.querySelector('form button[type=submit]')).toBeTruthy();
	});

	it('should render children if user is logged in and the component does not require a role', () => {
		jest.spyOn(useAuth, 'default').mockReturnValue({
			user: {
				email: 'test@test.com',
			},
		});

		const childrenText = 'children';

		const { container } = render(
			<Protected>
				<h1>{childrenText}</h1>
			</Protected>,
		);

		expect(container).toMatchSnapshot();
		expect(container.querySelector('h1').textContent).toEqual(childrenText);
	});

	it('should render children if user is logged in and has the role', () => {
		jest.spyOn(useAuth, 'default').mockReturnValue({
			user: {
				email: 'test@test.com',
				role: 'admin',
			},
		});

		const childrenText = 'children';

		const { container } = render(
			<Protected role="admin">
				<h1>{childrenText}</h1>
			</Protected>,
		);

		expect(container).toMatchSnapshot();
		expect(container.querySelector('h1').textContent).toEqual(childrenText);
	});

	it('should not render children if user is logged in and does not have the role', () => {
		jest.spyOn(useAuth, 'default').mockReturnValue({
			user: {
				email: 'test@test.com',
				role: 'moderator',
			},
		});

		const childrenText = 'children';

		const { container, getByTestId } = render(
			<Protected role="admin">
				<h1>{childrenText}</h1>
			</Protected>,
		);

		expect(container).toMatchSnapshot();
		expect(getByTestId('notAuthorized')).toBeInTheDocument();
		expect(container.querySelector('h1').textContent).not.toEqual(childrenText);
	});

	it('should redirect to another page when user is not logged in and component have a redirect prop', () => {
		jest.spyOn(Router, 'useRouter').mockReturnValue({
			push: jest.fn(),
		});

		const childrenText = 'children';
		const redirectTo = '/';

		const { container } = render(
			<Protected role="admin" redirectTo={redirectTo}>
				<h1>{childrenText}</h1>
			</Protected>,
		);

		expect(container).toMatchSnapshot();
		expect(container.querySelector('h1').textContent).not.toEqual(childrenText);
		expect(Router.useRouter().push).toHaveBeenCalledWith(redirectTo);
		expect(Router.useRouter().push).toHaveBeenCalledTimes(1);
	});

	it('should show an inline login when inline props was passed', () => {
		jest.spyOn(useAuth, 'default').mockReturnValue({
			user: {
				email: null,
			},
		});

		const childrenText = 'children';

		const { container, queryByTestId } = render(
			<Protected inline>
				<h1>{childrenText}</h1>
			</Protected>,
		);

		expect(container).toMatchSnapshot();
		expect(queryByTestId('modal')).toBeNull();
		expect(container.querySelector('form input[type=email]')).toBeTruthy();
		expect(container.querySelector('form input[type=password]')).toBeTruthy();
		expect(container.querySelector('form button[type=submit]')).toBeTruthy();
	});

	it('should return only the inline message when onlyUnauthorizedMessage is provided', () => {
		const childrenText = 'children';

		jest.spyOn(useAuth, 'default').mockReturnValue({
			user: {
				email: null,
			},
		});

		const { container, queryByTestId } = render(
			<Protected onlyUnauthorizedMessage>
				<h1>{childrenText}</h1>
			</Protected>,
		);

		expect(container).toMatchSnapshot();
		expect(container.querySelector('h1').textContent).not.toEqual(childrenText);
		expect(queryByTestId('modal')).not.toBeTruthy();
		expect(queryByTestId('notAuthorized')).toBeTruthy();
		expect(container.querySelector('form input[type=email]')).not.toBeTruthy();
		expect(container.querySelector('form input[type=password]')).not.toBeTruthy();
		expect(container.querySelector('form button[type=submit]')).not.toBeTruthy();
	});
});
