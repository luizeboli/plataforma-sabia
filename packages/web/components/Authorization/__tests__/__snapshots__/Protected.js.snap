// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Protected component should not render children if user is logged in and does not have the role 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  text-align: center;
  width: 100%;
  height: 100%;
  padding: 5rem 0;
}

.c0 h1 {
  color: hsla(0,0%,7%);
  font-size: 3.6rem;
  margin: 1.5rem auto;
}

<div>
  <div
    class="c0"
    data-testid="notAuthorized"
  >
    <h1>
      error:notAuthorized
    </h1>
  </div>
</div>
`;

exports[`Protected component should redirect to another page when user is not logged in and component have a redirect prop 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  text-align: center;
  width: 100%;
  height: 100%;
  padding: 5rem 0;
}

.c0 h1 {
  color: hsla(0,0%,7%);
  font-size: 3.6rem;
  margin: 1.5rem auto;
}

<div>
  <div
    class="c0"
    data-testid="notAuthorized"
  >
    <h1>
      error:notAuthorized
    </h1>
  </div>
</div>
`;

exports[`Protected component should render children if user is logged in and has the role 1`] = `
<div>
  <h1>
    children
  </h1>
</div>
`;

exports[`Protected component should render children if user is logged in and the component does not require a role 1`] = `
<div>
  <h1>
    children
  </h1>
</div>
`;

exports[`Protected component should return the login modal if user is not logged in 1`] = `
.c8 {
  font-size: 1.4rem;
  line-height: 14px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c8 > div.wrapper {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c9 {
  color: hsla(0,0%,40%);
  font-size: 1.4rem;
  font-weight: 500;
}

.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
}

.c11 {
  width: 100%;
  height: 4.4rem;
  font-size: 1.4rem;
  margin: 0.5rem 0;
  padding: 1.2rem;
  background: hsla(0,0%,100%);
  border: 1px solid hsla(0,0%,80%);
  border-radius: 0.2rem;
  color: hsla(0,0%,40%);
  opacity: 1;
}

.c11::-webkit-input-placeholder {
  color: hsla(0,0%,52%);
  font-weight: 300;
  font-style: italic;
}

.c11::-moz-placeholder {
  color: hsla(0,0%,52%);
  font-weight: 300;
  font-style: italic;
}

.c11:-ms-input-placeholder {
  color: hsla(0,0%,52%);
  font-weight: 300;
  font-style: italic;
}

.c11::placeholder {
  color: hsla(0,0%,52%);
  font-weight: 300;
  font-style: italic;
}

.c12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 1rem;
  font: 1.2rem;
}

.c13 {
  width: 2.5rem;
  height: 2.5rem;
  margin: 0rem;
  margin: 0 0.5rem 0 0;
}

.c14 {
  width: 100%;
  color: hsla(0,0%,40%);
}

.c7 {
  margin: 1rem auto;
  padding: 2rem 1rem;
  width: 100%;
}

.c7 button,
.c7 a {
  padding-right: 3rem;
  padding-left: 3rem;
}

.c7 > label {
  width: 100%;
  display: block;
  font-weight: 700;
  font-size: 1.7rem;
}

.c15 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c15 > a[href] {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  padding: 0 1rem;
}

.c15 > button {
  margin-top: 0.5rem;
  margin-left: 0rem;
}

.c6 {
  background-color: hsla(29,94%,62%);
  color: hsla(0,0%,100%);
  border-radius: 0.5rem;
  border: none;
  font-size: 2.2rem;
  text-transform: uppercase;
  padding: 1.8rem 6rem;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  display: inline-block;
}

.c6:hover {
  opacity: 0.8;
}

.c3 {
  width: 50rem;
}

.c3 button {
  width: 100%;
  background-color: hsla(29,94%,62%);
  margin-top: 2rem;
  margin-bottom: 2rem;
}

.c3 a {
  font-weight: 100;
}

.c5 {
  display: block;
  font-family: 'Museo',sans-serif;
  font-weight: 500;
  font-size: 3rem;
  color: hsla(0,0%,40%);
  padding-bottom: 2rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c4 {
  margin-top: 1rem;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c4 button {
  width: 100%;
  background-color: hsl(166,100%,34%);
  margin-top: 1.5rem;
  margin-bottom: 2rem;
}

.c16 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  color: hsla(29,94%,62%);
  font-weight: 500 !important;
  margin: 0rem !important;
  padding: 0rem !important;
}

.c0 {
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  z-index: 1500;
  background: rgba(1,1,1,0.5);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1 {
  position: fixed;
  min-width: 40rem;
  z-index: 2000;
  padding: 4rem;
  background: hsla(0,0%,100%);
  border-radius: 0.5rem;
  box-shadow: 0px 0px 20px -5px rgba(38,38,38,1);
}

.c2 {
  background: 0;
  border: 0;
  position: absolute;
  top: 1rem;
  right: 1rem;
}

.c2 svg {
  height: 3rem;
  width: 3rem;
  -webkit-transition: color 0.3s;
  transition: color 0.3s;
}

.c2 svg:hover {
  color: hsla(29,94%,62%);
}

.c17 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  text-align: center;
  width: 100%;
  height: 100%;
  padding: 5rem 0;
}

.c17 h1 {
  color: hsla(0,0%,7%);
  font-size: 3.6rem;
  margin: 1.5rem auto;
}

@media (max-width:768px) {
  .c15 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
  }

  .c15 > button {
    margin-left: 0;
  }
}

@media (max-width:768px) {
  .c6 {
    font-size: 1.6rem;
    padding: 1.4rem 6rem;
  }
}

@media (max-width:768px) {
  .c3 {
    width: 100%;
  }
}

@media (max-width:768px) {
  .c1 {
    min-width: 90%;
    padding: 2rem;
  }
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <button
        class="c2"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 1024 1024"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 0 1-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
          />
        </svg>
      </button>
      <div
        class="c3"
      >
        <div
          class="c4"
        >
          <div
            class="c5"
          >
            common:areYouNewHere?
          </div>
          <button
            class="c6"
            color="hsla(0, 0%, 100%)"
            title=""
            type="button"
          >
            common:registerYourUser
          </button>
        </div>
        <form
          class="c7"
        >
          <div
            class="c5"
          >
            common:alreadyRegistered?
          </div>
          <div
            class="c8"
          >
            <label
              class="c9"
              for="email"
            />
            <div
              class="c10"
            >
              <input
                aria-label=""
                aria-required="true"
                class="c11"
                id="email"
                name="email"
                placeholder="E-mail"
                type="email"
              />
            </div>
          </div>
          <div
            class="c8"
          >
            <label
              class="c9"
              for="password"
            />
            <div
              class="c10"
            >
              <input
                aria-label=""
                aria-required="true"
                class="c11"
                id="password"
                name="password"
                placeholder="Senha"
                type="password"
              />
            </div>
          </div>
          <div
            class="c12"
          >
            <input
              aria-label="common:rememberpassword"
              aria-required="false"
              class="c13"
              id="remember"
              name="remember"
              type="checkbox"
            />
            <div
              class="c14"
            >
              common:rememberpassword
            </div>
          </div>
          <p>
            common:signInToContinue
          </p>
          <div
            class="c15"
          >
            <button
              class="c6"
              color="hsla(0, 0%, 100%)"
              title=""
              type="submit"
            >
              common:login
            </button>
            <a
              class="c16"
            >
              common:forgotPassword
            </a>
          </div>
        </form>
      </div>
    </div>
  </div>
  <div
    class="c17"
    data-testid="notAuthorized"
  >
    <h1>
      error:notAuthorized
    </h1>
  </div>
</div>
`;
